controlplane/kubeadm/internal/filters.go

```diff
@@ -29,101 +31,136 @@ import (
+
+	if !collections.MatchesKubernetesVersion(kcp.Spec.Version)(machine) {
+		machineVersion := ""
+		if machine != nil && machine.Spec.Version != nil {
+			machineVersion = *machine.Spec.Version
+		}
+		mismatchReasons = append(mismatchReasons, fmt.Sprintf("Machine version %q is not equal to KCP version %q", machineVersion, kcp.Spec.Version))
+	}
+
+	if reason, matches := matchesKubeadmBootstrapConfig(machineConfigs, kcp, machine); !matches {
+		mismatchReasons = append(mismatchReasons, reason)
+	}
+
+	if reason, matches := matchesTemplateClonedFrom(infraConfigs, kcp, machine); !matches {
+		mismatchReasons = append(mismatchReasons, reason)
+	}
+
+	if len(mismatchReasons) > 0 {
+		return fmt.Sprintf("Machine does not match current spec: %s", strings.Join(mismatchReasons, ",")), false
+	}
+
+	return "", true
 }
 
 // NeedsRollout returns a filter to determine if a machine needs rollout.
-func NeedsRollout(reconciliationTime, rolloutAfter *metav1.Time, rolloutBefore *controlplanev1.RolloutBefore, infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane) func(machine *clusterv1.Machine) bool {
-	return collections.Or(
-		// Machines whose certificates are about to expire.
-		collections.ShouldRolloutBefore(reconciliationTime, rolloutBefore),
-		// Machines that are scheduled for rollout (KCP.Spec.RolloutAfter set, the RolloutAfter deadline is expired, and the machine was created before the deadline).
-		collections.ShouldRolloutAfter(reconciliationTime, rolloutAfter),
-		// Machines that do not match with KCP config.
-		collections.Not(MatchesMachineSpec(infraConfigs, machineConfigs, kcp)),
-	)
+func NeedsRollout(reconciliationTime, rolloutAfter *metav1.Time, rolloutBefore *controlplanev1.RolloutBefore, infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (string, bool) {
+	rolloutReasons := []string{}
+
+	// Machines whose certificates are about to expire.
+	if collections.ShouldRolloutBefore(reconciliationTime, rolloutBefore)(machine) {
+		rolloutReasons = append(rolloutReasons, "rolloutBefore requires a rollout: certificates will expire soon")
+	}
+
+	// Machines that are scheduled for rollout (KCP.Spec.RolloutAfter set,
+	// the RolloutAfter deadline is expired, and the machine was created before the deadline).
+	if collections.ShouldRolloutAfter(reconciliationTime, rolloutAfter)(machine) {
+		rolloutReasons = append(rolloutReasons, "rolloutAfter requires a rollout")
+	}
+
+	// Machines that do not match with KCP config.
+	if mismatchReason, matches := matchesMachineSpec(infraConfigs, machineConfigs, kcp, machine); !matches {
+		rolloutReasons = append(rolloutReasons, mismatchReason)
+	}
+
+	if len(rolloutReasons) > 0 {
+		return fmt.Sprintf("Machine %s needs rollout because: %s", machine.Name, strings.Join(rolloutReasons, ",")), true
+	}
```

Comment:
Same as above
```suggestion
	if len(rolloutReasons) > 0 {
		return strings.Join(rolloutReasons, ","), true
	}
```

Reply:
If I apply this change and the one above instead of

> Rolling out Control Plane machines: Machine clusterclass-rollout-vtjq3t-2w8vj-62x9w needs rollout because: Machine does not match current spec: Machine InitConfiguration or JoinConfiguration do not match the ones on the KCP object

we get

> Rolling out Control Plane machines: Machine InitConfiguration or JoinConfiguration are outdated

Which means we don't know anymore to which Machine this reason belongs